Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java

```diff
@@ -339,35 +352,79 @@ private DetectAction(ConcurrentLinkedQueue<MovecraftLocation> currentFrontier, C
         @Override
         public void run() {
             MovecraftLocation probe;
+
             while((probe = currentFrontier.poll()) != null) {
-                visitedMaterials.computeIfAbsent(movecraftWorld.getMaterial(probe), Functions.forSupplier(ConcurrentLinkedDeque::new)).add(probe);
+                BlockData bd = movecraftWorld.getData(probe);
+                Material m = bd.getMaterial();
+                boolean blockFacingCraft = true;
+
+                if (m == Material.LADDER) {
+                    BlockFace facing = ((Directional) bd).getFacing().getOppositeFace();
+                    MovecraftLocation relativeLoc = probe.getRelative(facing);
+
+                    if (!legal.contains(relativeLoc)) {
+                        blockFacingCraft = false; // Invalidate block if it's not facing the craft
+                    }
+                } else if (bd instanceof FaceAttachable attachable) {
+                    FaceAttachable.AttachedFace attachedFace = attachable.getAttachedFace();
+                    BlockFace facing = toBlockFace(attachedFace, bd);
+                    MovecraftLocation relativeLoc = probe.getRelative(facing);
+
+                    if (!legal.contains(relativeLoc)) {
+                        blockFacingCraft = false; // Invalidate block if it's not facing the craft
+                    }
+                } else if (bd instanceof Lantern lantern) {
+                    BlockFace facing = toBlockFace(lantern);
+                    MovecraftLocation relativeLoc = probe.getRelative(facing);
+
+                    if (!legal.contains(relativeLoc)) {
+                        blockFacingCraft = false; // Invalidate block if it's not facing the craft
+                    }
+                }
+
+                if (!blockFacingCraft) {
```

Comment:
Remove this if statement and inline the `continue` to every line you do `blockFacingCraft=false`, than remove `blockFacingCraft`